keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -40,74 +56,145 @@ For operators who want to regularly rotate the signing and verifying keys for pr
+As of Kubernetes `v1.13.2`, The Kubernetes Control Plane config accepts `ServiceAccountIssuer` which is an interface of type `serviceaccount.TokenGenerator` as defined in `pkg/serviceaccount/jwt.go`. This enables defining alternate implementations of token generators. Thus, the default TokenGenerator can be replaced by one that integrates with KeyService.
 
-### New API
+### Updates for supported public keys
 
-I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+Starting ` v1.31` (via. the [PR](https://github.com/kubernetes/kubernetes/pull/125177) by liggit@), `ServiceAccountPublicKeys` in ApiServer config was replaced by `ServiceAccountPublicKeysGetter` interface. The static keys held by `ServiceAccountPublicKeys` were hence moved to be managed by the implementation called `StaticPublicKeysGetter`. This presents an opportunity to add a new implementation, one that will get public keys from KeyService depending on how the cluster is configured.
 
-```proto
-syntax = "proto3";
+In the same PR, `OpenIDMetadataServer` was updated to accept implementation of `OpenIDMetadataProvider` as defined in  `pkg/serviceaccount/openidmetadata.go`. `OpenIDMetadataProvider` interacts with `ServiceAccountPublicKeysGetter` for serving supported keys which enables flexible replacing `StaticPublicKeysGetter` with a dynamic key getter (will integrate with KeyService) depending on how cluster is configured.
+
+### Notes/Constraints/Caveats
+
+- A new flag `--key-service-url` will be added to the API server specifying a unix socket where the key service will be accessible.
+- cumulation on local and external keys will not be supported.
+- If Both, `--service-account-key-file` and `--key-service-url` are specified, keys from external signers will override the keys at `service-account-key-file` as supported public keys.
+
+### Risks and Mitigations
 
-package v1alpha1;
+- New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.
+- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help ApiServer control the payload and thus the claims.
 
+## Design Details
+
+### KeyService RPC
+
+will be served on a socket as configured via the `--key-service-url` flag.
+
+```proto
+// This service is served by a process on a local Unix Domain Socket.
 service KeyService {
```

Comment:
KeyService is very generic, and could also describe the KMS plugin for etcd storage.  We should use a more specific name, like "KubernetesExternalJWTSigner", or something like that.

Reply:
calling it `ExternalJWTSigner`